sktime/clustering/_time_series_k_means.py

```diff
@@ -0,0 +1,159 @@
+
+    __averaging_algorithm_dict: Averaging_Algo_Dict = {
+        "dba": BarycenterAveraging,
+        "mean": MeanAveraging,
+    }
+
+    def __init__(
+        self,
+        n_clusters: int = 8,
+        init_algorithm: Init_Algo = "random",
+        max_iter: int = 300,
+        verbose: bool = False,
+        metric: Metric_Parameter = "dtw",
+        averaging_algorithm: Averaging_Algo = "mean",
+        averaging_algorithm_iterations: int = 50,
+    ):
+        """Constructor for TimeSeiresKMeans clusterer
+
+        Parameters
+        ----------
+            n_clusters: int, default = 8
+                The number of clusters to form as the number of
+                centroids to generate.
+
+            n_init: int, default = 10
+                Number of time the k-means algorithm will be run
+                with different centroid seeds. The final result
+                will be the best output of n_init consecutive runs
+                in terms of inertia.
+
+            init_algorithm: str, default = random
+                Algorithm that is used to initialise the cluster
+                centers. See clustering/base/base_types.py
+                for description of type. The following are the
+                str supported types:
+                'random' = random initialisation
+
+            max_iter: int, default = 300
+                Maximum number of iterations of time series k means
+                for a single run.
+
+            verbose: bool, default = False
+                Verbosity mode.
+
+            metric: Metric_Parameter, default = None
+                The distance metric that is used to calculate the
+                distance between points.
+
+            averaging_algorithm: Averaging_Algo
+                The method used to create the average from a cluster
+
+            averaging_algorithm_iterations: int
+                Where appropriate (i.e. DBA) the average is refined by
+                iterations. This is the number of times it is refined
+        """
+        super(TimeSeriesKMeans, self).__init__(
+            n_clusters=n_clusters,
+            init_algorithm=init_algorithm,
+            max_iter=max_iter,
+            verbose=verbose,
+            metric=metric,
+        )
+        metric_str = None
+        if isinstance(metric, str):
+            metric_str = metric
+
+        if isinstance(averaging_algorithm, str):
+            if metric_str is not None and averaging_algorithm == "auto":
+                if metric_str == "dtw":
+                    averaging_algorithm = "dba"
+                else:
+                    averaging_algorithm = "mean"
+
+            self.averaging_algorithm = TimeSeriesKMeans.__averaging_algorithm_dict[
+                averaging_algorithm
+            ]
+        else:
+            self.averaging_algorithm = averaging_algorithm
+        self.averaging_algorithm_iterations = averaging_algorithm_iterations
+
+    def fit(self, X: Data_Frame) -> None:
```

Comment:
Question: Should we follow the same interface as for classification as in scikit-learn, keeping `y` in the signature but ignoring it? Not sure if that's necessary for re-using some of the higher-level composition functionality. @fkiraly any idea?

Reply:
We would want to accept 3d numpy arrays as they seem to be much easier to understand and work with than the nested data frame and we have all the functionality for input validation/conversions if required. 
```suggestion
    def fit(self, X: Panel) -> None:
```
where `Panel = Union[pd.DataFrame, np.ndarray]` (e.g. 3d or nested array), as in `sktime/transformations/base.py`